plutus-core/plutus-ir/src/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -400,3 +401,16 @@ withTyVarsOfBindings = flip $ foldr withTyVarsOfBinding
 -- | Helper to add type variables into a computation's environment.
 withTyVarDecls :: [TyVarDecl TyName ann] -> TypeCheckM uni fun c e a -> TypeCheckM uni fun c e a
 withTyVarDecls = flip . foldr $ \(TyVarDecl _ n k) -> withTyVar n $ void k
+
+-- | Substitute `TypeBind`s from the given list of `Binding`s in the given `Type`.
+-- This is so that @let a = (con integer) in \(x : a) -> x@ typechecks.
+substTypeBinds ::
+    HasUniApply uni =>
+    NonEmpty (Binding TyName Name uni fun ann) ->
+    Normalized (Type TyName uni ()) ->
+    PirTCEnv uni fun e (Normalized (Type TyName uni ()))
+substTypeBinds = flip . foldrM $ \b ty -> case b of
+    TypeBind _ tvar rhs -> do
+        rhs' <- normalizeTypeM (void rhs)
+        substNormalizeTypeM rhs' (tvar ^. tyVarDeclName) (unNormalized ty)
```

Comment:
Hm, I'm actually surprised that `substNormalizeTypeM` doesn't take a normalized type, I wonder if it should (@effectfully ?).

I also wonder if it would be worth writing a version that does all the substitutions before normalizing. Probably doesn't matter in practice.

Reply:
There's a note about that:

```
{- Note [Normalizing substitution]
@substituteNormalize[M]@ is only ever used as normalizing substitution that receives two already
normalized types. However we do not enforce this in the type signature, because
1) it's perfectly correct for the last argument to be non-normalized
2) it would be annoying to wrap 'Type's into 'NormalizedType's
-}
```

Check how the function is used in `normalizeTypeM` and `inferTypeM` to see what I meant by 2.

@zliu41 while you're here, could you replace `substituteNormalize[M]` in that Note with `substNormalizeTypeM` (slipped through some refactoring I guess)?